lib/epiviewpoint/data_file.ex

```diff
@@ -0,0 +1,86 @@
+defmodule EpiViewpoint.DataFile do
+  alias Explorer.DataFrame
+  alias EpiViewpoint.Extra
+
+  NimbleCSV.define(EpiViewpoint.DataFile.Parser,
+    separator: ",",
+    escape: "\"",
+    line_separator: "\r\n",
+    trim_bom: true,
+    moduledoc: """
+    A CSV parser that uses comma as separator and double-quotes as escape according to RFC4180,
+    and trims byte-order marks (BOMs) that may be generated by some systems.
+    """
+  )
+
+  def read(string, :csv, header_transformer, headers) when is_binary(string) do
+    read(string, header_transformer, headers, &parse_csv/1)
+  end
+
+  def read(string, :ndjson, header_transformer, headers) when is_binary(string) do
+    read(string, header_transformer, headers, &parse_ndjson/1)
+  end
+
+  def read(input, header_transformer, [required: required_headers, optional: optional_headers], parser) do
+    with {:ok, df} <- parser.(input),
+         provided_headers <- df |> DataFrame.names() |> Enum.map(&String.trim/1) |> header_transformer.(),
```

Comment:
Extract to a private function. I prefer for `with` to be clean and readable.

Reply:
sorry I missed this, will update soon

Reply:
updated.